src/vmm/src/guest_config/cpuid/normalize.rs

```diff
@@ -50,6 +49,80 @@ pub enum GetMaxCpusPerPackageError {
     Overflow,
 }
 
+/// Error type for setting a bit range.
+#[derive(Debug, PartialEq, Eq, thiserror::Error)]
+#[error("Given value is greater than maximum storable value in bit range.")]
+pub struct CheckedAssignError;
+
+/// Sets a given bit to a true or false (1 or 0).
+#[allow(clippy::integer_arithmetic, clippy::arithmetic_side_effects)]
+pub fn set_bit(x: &mut u32, bit: u8, y: bool) {
+    debug_assert!(bit < 32);
+    *x = (*x & !(1 << bit)) | ((u32::from(u8::from(y))) << bit);
+}
+
+/// Sets a given range to a given value.
+#[allow(clippy::integer_arithmetic, clippy::arithmetic_side_effects)]
+pub fn set_range(
+    x: &mut u32,
+    range: std::ops::Range<u8>,
+    y: u32,
+) -> Result<(), CheckedAssignError> {
+    debug_assert!(range.end >= range.start);
+    match range.end - range.start {
+        z @ 0..=31 => {
+            if y >= 2u32.pow(u32::from(z)) {
+                Err(CheckedAssignError)
+            } else {
+                let shift = y << range.start;
+                *x = shift | (*x & !mask(range));
+                Ok(())
+            }
+        }
+        32 => {
+            let shift = y << range.start;
+            *x = shift | (*x & !mask(range));
+            Ok(())
+        }
+        33.. => Err(CheckedAssignError),
+    }
+}
+
+/// Returns a mask where the given range is ones.
+#[allow(
+    clippy::as_conversions,
+    clippy::integer_arithmetic,
+    clippy::arithmetic_side_effects,
+    clippy::cast_possible_truncation
+)]
+const fn mask(range: std::ops::Range<u8>) -> u32 {
+    debug_assert!(range.end >= range.start);
+    let size = 8 * std::mem::size_of::<u32>();
+    debug_assert!(range.end as usize <= size);
+
+    let front = shift(range.start, size as u8);
+    let back = shift(range.end, size as u8);
+    !front & back
+}
+/// Returns a value where in the binary representation all bits to the right of the x'th bit from
+/// the left are 1.
+#[allow(
+    clippy::integer_arithmetic,
+    clippy::arithmetic_side_effects,
+    clippy::unreachable
+)]
+const fn shift(x: u8, max: u8) -> u32 {
```

Comment:
Shouldn't here be another `debug_assert!(max < 32)`? Otherwise you can call it with `shift(100, 200)`, but it does not make sense.
Or maybe just add it as a lambda inside `mask` method?

Reply:
nit: Maybe a little refactoring might be useful. Like removing `unreachable` + `shift` technically does not need second arg, because it is always 32.

Reply:
Ok, this version is looks good to me too.